tests/integration/datastore/test_datastore.py

```diff
@@ -14,64 +14,392 @@
+        rec.bar = b'two record'
+
+    # Let's query!
+    with storage.query_by(FooRecord) as records:
+        assert len(records) == 3
+        assert type(records) == list
+        assert records[0]._DatastoreRecord__writeable is False
+        assert records[1]._DatastoreRecord__writeable is False
+        assert records[2]._DatastoreRecord__writeable is False
+
+    # Try with BarRecord
+    with storage.query_by(BarRecord) as records:
+        assert len(records) == 2
+
+    # Try to query for non-existent records
+    with pytest.raises(datastore.RecordNotFound):
+        with storage.query_by(NoRecord) as records:
+            assert len(records) == 'this never gets executed cause it raises'
+
+    # Queries without writeable are read only
+    with pytest.raises(datastore.DatastoreTransactionError):
+        with storage.query_by(FooRecord) as records:
+            records[0].foo = b'this should error'
+
+    # Let's query by specific record and field
+    with storage.query_by(BarRecord, filter_field='foo') as records:
+        assert len(records) == 1
+
+    # Query for a non-existent field in an existing record
+    with pytest.raises(datastore.RecordNotFound):
+        with storage.query_by(FooRecord, filter_field='bar') as records:
+            assert len(records) == 'this never gets executed cause it raises'
+
+    # Query for a non-existent field that is _similar to an existing field_
+    with pytest.raises(datastore.RecordNotFound):
+        with storage.query_by(FooRecord, filter_field='fo') as records:
+            assert len(records) == 'this never gets executed cause it raises'
+
+    # Query for a field with a filtering function
+    # When querying with a field _and_ a filtering function, the `filter_func`
+    # callable is given the field value you specified.
+    # We throw a `isinstance` in there to ensure that the type given is a field value and not a record
+    filter_func = lambda field_val: not isinstance(field_val, DatastoreRecord) and field_val == b'another record'
+    with storage.query_by(FooRecord, filter_field='foo', filter_func=filter_func) as records:
+        assert len(records) == 2
+        assert records[0].foo == b'another record'
+        assert records[1].foo == b'another record'
+
+    # Query with _only_ a filter func.
+    # This filter_func will receive a `DatastoreRecord` instance that is readonly
+    filter_func = lambda field_rec: isinstance(field_rec, DatastoreRecord) and field_rec.foo == b'one record'
+    with storage.query_by(FooRecord, filter_func=filter_func) as records:
+        assert len(records) == 1
+        assert records[0].foo == b'one record'
+
+        # This record isn't writeable
+        with pytest.raises(TypeError):
+            records[0].foo = b'this will error'
+
+    # Make a writeable query on BarRecord
+    with storage.query_by(BarRecord, writeable=True) as records:
+        records[0].bar = b'this writes'
+        records[1].bar = b'this writes'
```

Comment:
These values aren't actually written to the datastore until the context manager is exited..correct? If so then the assertions below that check the written value should instead be done outside of this context manager, in a different one - I guess the records would have to be queried for again, and then the values checked

Reply:
Technically, none of the statements in a context manager are executed until it's exited and all exceptions are handled, I believe.

So you can do either/or: check records in the context manager, or check records outside of the context manager -- both work fine. When I do it inside/outside, I just do it to make the action being performed obvious, e.g. a query for a record with a field that didn't get written is a `RecordNotFound`, etc.

Reply:
I reread what I wrote here, and I'm not very keen on my answer.

You shouldn't confuse the write `records[0].bar = b'this writes'` with something like a buffer. When that statement is executed, `b'this writes'` is written to lmdb. What lmdb does behind the scenes is ensure that all errors are handled before being written to disk, I think.

Technically, yes, it's not written until the context manager is exited. When the database transaction is finalized (I think on that context manager exit), it goes from being written to memory to being written to the disk. So, while you're in the context manager, you can make these read statements and they will read from memory with what lmdb intends to write to disk.

You point out that the check should be outside the context manager to truly assert a read -- and you're right. If you really want to confirm a write, it should be outside the context manager, but for the purposes of some of these tests it doesn't seem very important to assert that. What I'm really trying to test is the `Datastore` and not lmdb.

Reply:
👍